[3124] Custom Resources rewrite (#3185)
chef/chef-web-docs: content/custom_resources/_index.md

    identifier = "chef_infra/cookbook_reference/custom_resources/custom_resources Custom Resources"
    parent = "chef_infra/cookbook_reference/custom_resources"
    weight = 10
+++

A custom resource:

- Is a simple extension of Chef Infra Client that adds your own resources
- Is implemented and shipped as part of a cookbook
- Leverages resources that are built into Chef Infra Client and/or custom Ruby code
- Is reusable in the same way as resources that are built into Chef Infra Client

For example, Chef Infra Client includes built-in resources to manage
files, packages, templates, and services, but it does not include a
resource that manages websites.

## Alternative Reading

Learn Chef interactive tutorial - [Extending Chef Infra: Custom Resources](https://learn.chef.io/courses/course-v1:chef+Infra201+Perpetual/about)

## Syntax

A custom resource is defined as a Ruby file and is located in a
cookbook's `/resources` directory. This file:

- Declares the properties of the custom resource
- Defines actions the custom resource may take

The layout for a custom resource is:

```ruby
property :property_name, RubyType, default: 'value'

action :action_a do

>  # a mix of built-in Chef resources and Ruby

Reviewer: ```suggestion
 # a mix of built-in Chef Infra resources and Ruby
```

end

action :action_b do

>  # a mix of built-in Chef resources and Ruby

Reviewer: ```suggestion
 # a mix of built-in Chef Infra resources and Ruby
```

end
```

The first action listed is the default action.

### Example Resource

> This example `site` utilizes Chef's built-in `file`, `service` and

Reviewer: ```suggestion
This example `site` utilizes Chef Infra's built-in `file`, `service` and
```
